src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -1255,25 +1255,36 @@ internal virtual void ApplyMappedFileChanges(SolutionChanges solutionChanges)
 
         private void CheckAllowedSolutionChanges(SolutionChanges solutionChanges)
         {
-            if (solutionChanges.GetRemovedProjects().Any() && !this.CanApplyChange(ApplyChangesKind.RemoveProject))
+            if (!CanApplyChange(ApplyChangesKind.RemoveProject) && solutionChanges.GetRemovedProjects().Any())
             {
                 throw new NotSupportedException(WorkspacesResources.Removing_projects_is_not_supported);
             }
 
-            if (solutionChanges.GetAddedProjects().Any() && !this.CanApplyChange(ApplyChangesKind.AddProject))
+            if (!CanApplyChange(ApplyChangesKind.AddProject) && solutionChanges.GetAddedProjects().Any())
             {
                 throw new NotSupportedException(WorkspacesResources.Adding_projects_is_not_supported);
             }
 
+            if (!CanApplyChange(ApplyChangesKind.AddSolutionAnalyzerReference) && solutionChanges.GetAddedAnalyzerReferences().Any())
+            {
+                throw new NotSupportedException(WorkspacesResources.Adding_analyzer_references_is_not_supported);
+            }
+
+            if (!CanApplyChange(ApplyChangesKind.RemoveSolutionAnalyzerReference) && solutionChanges.GetRemovedAnalyzerReferences().Any())
+            {
+                throw new NotSupportedException(WorkspacesResources.Removing_analyzer_references_is_not_supported);
+            }
+
             foreach (var projectChanges in solutionChanges.GetProjectChanges())
             {
-                this.CheckAllowedProjectChanges(projectChanges);
+                CheckAllowedProjectChanges(projectChanges);
             }
         }
 
         private void CheckAllowedProjectChanges(ProjectChanges projectChanges)
         {
-            if (projectChanges.OldProject.CompilationOptions != projectChanges.NewProject.CompilationOptions)
+            if (!CanApplyChange(ApplyChangesKind.ChangeCompilationOptions) &&
```

Comment:
I guess this is correct but I admit it threw me for a loop and this might be worthy of a comment. (And I see I also need to merge https://github.com/dotnet/roslyn/pull/45243...). Our contract here is a bit funky in that you are supposed to return _false_ for CanApplyChange if you want to be called for the more explicit questions -- the legacy method just means you opt-in to all changes, no matter what they are.